tracing-appender/src/rolling.rs

```diff
@@ -0,0 +1,178 @@
+        Self { inner }
+    }
+
+    fn inner(&self) -> io::Result<MutexGuard<'a, InnerAppender<BufWriterFactory>>> {
+        self.inner.lock().map_err(
+            |_: std::sync::PoisonError<MutexGuard<InnerAppender<BufWriterFactory>>>| {
+                io::Error::from(io::ErrorKind::Other)
+            },
+        )
+    }
+}
+
+impl<'a> io::Write for RollingFileWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner()?.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner()?.flush()
+    }
+}
+
+impl<'a> MakeWriter for RollingFileWriter<'a> {
+    type Writer = RollingFileWriter<'a>;
+
+    fn make_writer(&self) -> Self::Writer {
+        RollingFileWriter::new(&self.inner)
+    }
+}
+
+pub fn hourly<P: AsRef<Path>>(directory: P, file_name_prefix: P) -> RollingFileAppender {
+    create_writer(
+        Rotation::HOURLY,
+        directory.as_ref(),
+        file_name_prefix.as_ref(),
+    )
+}
+
+pub fn daily<P: AsRef<Path>>(directory: P, file_name_prefix: P) -> RollingFileAppender {
+    create_writer(
+        Rotation::DAILY,
+        directory.as_ref(),
+        file_name_prefix.as_ref(),
+    )
+}
+
+pub fn never<P: AsRef<Path>>(directory: P, file_name_prefix: P) -> RollingFileAppender {
+    create_writer(
+        Rotation::NEVER,
+        directory.as_ref(),
+        file_name_prefix.as_ref(),
+    )
+}
+
+fn create_writer(
+    rotation: Rotation,
+    directory: &Path,
+    file_name_prefix: &Path,
+) -> RollingFileAppender {
+    RollingFileAppender {
+        inner: Mutex::new(
+            InnerAppender::new(
+                directory,
+                file_name_prefix,
+                rotation,
+                BufWriterFactory {},
+                Utc::now(),
+            )
+            .expect("Failed to create appender"),
+        ),
+    }
+}
+
+pub(crate) trait WriterFactory: Debug + Clone + Send {
+    type W: Write + Debug + Send;
+
+    fn create_writer(&self, directory: &str, filename: &str) -> io::Result<Self::W>;
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct BufWriterFactory {}
+
+impl WriterFactory for BufWriterFactory {
+    type W = BufWriter<File>;
+
+    fn create_writer(&self, directory: &str, filename: &str) -> io::Result<BufWriter<File>> {
+        let file_path = Path::new(directory).join(filename);
+        Ok(BufWriter::new(open_file_create_parent_dirs(&file_path)?))
+    }
+}
+
+// Open a file - if it throws any error, try creating the parent directory and then the file.
+fn open_file_create_parent_dirs(path: &Path) -> io::Result<File> {
+    let new_file = OpenOptions::new().append(true).create(true).open(path);
+    if new_file.is_err() {
+        if let Some(parent) = path.parent() {
+            fs::create_dir_all(parent)?;
+            return OpenOptions::new().append(true).create(true).open(path);
+        }
+    }
+
+    new_file
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub(crate) struct Rotation(RotationKind);
```

Comment:
If this is no longer part of the public API, we don't need to do the enum trick.

Reply:
Since we'll have RollingFileAppender's constructor public, this enum would still be part of the public API.